ktor-client/ktor-client-core/common/src/io/ktor/client/statement/HttpStatement.kt

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2014-2019 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.statement
+
+import io.ktor.client.*
+import io.ktor.client.call.*
+import io.ktor.client.request.*
+import io.ktor.http.*
+import io.ktor.http.cio.websocket.*
+import io.ktor.util.*
+import io.ktor.utils.io.*
+import io.ktor.utils.io.charsets.*
+import io.ktor.utils.io.core.*
+import kotlinx.coroutines.*
+import kotlin.reflect.*
+
+/**
+ * Prepared statement for http client request.
+ * This statement doesn't perform any network requests until [execute] method call.
+ *
+ * [HttpStatement] is safe to execute multiple times.
+ */
+class HttpStatement(
+    private val builder: HttpRequestBuilder,
+    private val client: HttpClient
+) {
+    /**
+     * Executes this statement and call the [block] with the streaming [response].
+     *
+     * The [response] argument holds a network connection until the [block] isn't completed. You can read the body
+     * on-demand or at once with [receive<T>()] method.
+     *
+     * After [block] finishes, [response] will be completed body will be discarded or released depends on the engine configuration.
+     */
+    suspend fun <T> execute(block: suspend (response: HttpResponse) -> T): T {
+        val response: HttpResponse = executeUnsafe()
+
+        try {
+            return block(response)
+        } finally {
+            val job = response.coroutineContext[Job]!! as CompletableJob
+
+            job.apply {
+                complete()
+                try {
+                    response.content.cancel()
+                } catch (_: Throwable) {}
```

Comment:
So what kind of errors do we suppress here?

Reply:
Already canceled

Reply:
Cancelling already cancelled should have no effect and should be no exception, isn't it?